stacks-signer/src/v0/stackerdb.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2024 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+use blockstack_lib::net::api::poststackerdbchunk::StackerDBErrorCodes;
+use hashbrown::HashMap;
+use libsigner::v0::messages::{MessageSlotID, SignerMessage};
+use libsigner::{SignerSession, StackerDBSession};
+use libstackerdb::{StackerDBChunkAckData, StackerDBChunkData};
+use slog::{slog_debug, slog_warn};
+use stacks_common::codec::StacksMessageCodec;
+use stacks_common::types::chainstate::StacksPrivateKey;
+use stacks_common::{debug, warn};
+
+use crate::client::{retry_with_exponential_backoff, ClientError, SignerSlotID};
+use crate::config::SignerConfig;
+
+/// The StackerDB client for communicating with the .signers contract
+#[derive(Debug)]
+pub struct StackerDB {
```

Comment:
There's quite a lot of code shared between v0 and v1's StackerDB (ie same struct, `new()`, `send_message_with_retry`). Would it be better to have a shared "base" StackerDB, and then version-specific functions that use it (like `get_encrypted_signer_state` in v1)? I'm pretty open to exactly how that would look, but I think it would be an improvement to have more shared code.

Reply:
It is not actually the same struct. v1 has next_transactions_session which is not a concept in v0. I can wrap the underlying similar stuff, but it seems a bit overkill since there is minimal overlap between the two especially since they use different definitions of MessageSlotID. I would have to make these generic in some way.

Reply:
K I added some traits and made it work.

Reply:
Ty! Looks good